### DISCUSSION TITLE
[Q&A] Ratchet vs Candle

### DISCUSSION THREAD

**Discussion by DavidGOrtega** [huggingface/ratchet#169] [General]
@FL33TW00D I know ratchet prior to be a HF project (congrats to you for joining 🤗 !!!) however the question that I have is how this project relates to Candle. To me, they have pretty common objectives.

**Comment by FL33TW00D** [comment]
Hi @DavidGOrtega,
There are a few key differences between Candle & Ratchet:
1. We are focused entirely on inference with WebGPU, and as they say _"through specialisation comes efficiency"_.
2. Huge focus on being cross platform. If we don't run with GPU acceleration out of the box on Windows, MacOS and Linux, that's a bug. I think right now we are one of the only frameworks with DirectX support, for speedy inference on Windows.
3. We don't want to just run the model for you. We want to build a toolkit that makes building web applications with AI integrated really easy. This means managing model caching, auto fetching from the hub, versioning, providing tooling for audio & video etc. etc.
4. We have **0 support for training**, unlike Candle, which allows us to make some optimisations that aren't currently done in Candle.

More than the above, we wanted to explore a different part of the design space, which can be easier separately. We are hoping one day to merge the best ideas from Candle & Ratchet 🤝. There are tons of places where Candle vastly surpasses us.

**Reply by EricLBuehler** [reply]
@FL33TW00D, thanks for your great work here! What optimizations can you make that are not currently done in Candle because you have no support for training?

**Reply by FL33TW00D** [reply]
Hi @EricLBuehler,

The 3 main optimisations are inplacing, our memory allocator, and our lazy evaluation.
<img width="975" alt="image" src="https://github.com/huggingface/ratchet/assets/45471420/0caaabab-fcd9-47e3-9a0e-b48f40064509">

Boxes outlined in red are operations where we automatically mutate the input buffer iff there are no further uses of the input buffer. This is super important for speed. In the future when we have a JIT, these won't even mutate the input buffer, the kernels will just be fused.

You can't do this if you need to train as you might need the buffers for the backward.

If you don't have lazy eval, then you have to submit work to the GPU after N amount of operations.
We submit work to the GPU in huge chunks. We can discuss on discord if you want to know more 🖖🏻

**Reply by EricLBuehler** [reply]
That sounds exciting, I'm excited to try out the JIT kernel fusion! Do you know if llama.cpp does a similar thing (as in building a graph and then applying these optimizations)?

**Reply by FL33TW00D** [reply]
@EricLBuehler llama.cpp is also lazy! MLX, tinygrad & llama.cpp are all lazy. Work graphs are becoming a big thing and I think lazy is the future (also known as unified execution from [this paper](http://scis.scichina.com/en/2020/222103.pdf))

Unsure if `llama.cpp` does the inplacing etc, but I would assume so for speed.

JIT Kernel fusion will be a big piece of work, hoping we get around to it sooner rather than later!

**Reply by EricLBuehler** [reply]
That is good to know! Do you have any benchmarks (even qualitative) of how lazy evaluation improved Ratchet's performance, or was Ratchet designed to be lazily evaluated from the ground up?

**Reply by FL33TW00D** [reply]
@EricLBuehler Lazy from day 0! You can disable the `inplace` operations with a flag to see the impact they have (it's substantial).

**Reply by EricLBuehler** [reply]
Great! Do you have any numbers (1x, 1.5x)?

**Reply by FL33TW00D** [reply]
@EricLBuehler Not really unfortunately, I don't have many `non-inplace` kernels, as I don't really need them.
I would expect its ~20%.

**Reply by EricLBuehler** [reply]
Nice, excellent work here!

**Comment by DavidGOrtega** [comment]
> Huge focus on being cross platform. If we don't run with GPU acceleration out of the box on Windows, MacOS and Linux, that's a bug. I think right now we are one of the only frameworks with DirectX support, for speedy inference on Windows.

I also think that going with wgpu is the way to go. However thats seems to be also [Candle's objetive ](https://github.com/huggingface/candle/issues/344). 
Could be Ratchet the missing backend?

**Reply by FL33TW00D** [reply]
I think yes one day the work from Ratchet could end up as a backend for Candle!
